core/node/base_token_adjuster/src/base_token_adjuster.rs

```diff
@@ -0,0 +1,429 @@
+
+            let result = conn.base_token_dal().get_latest_price().await;
+
+            drop(conn);
+
+            if let Ok(last_storage_price) = result {
+                return Ok(last_storage_price
+                    .base_token_price
+                    .div(&last_storage_price.eth_price));
+            } else {
+                if attempts >= max_retries {
+                    break;
+                }
+                let sleep_duration = Duration::from_secs(retry_delay)
+                    .mul_f32(rand::thread_rng().gen_range(0.8..1.2));
+                tracing::warn!(
+                    "Attempt {}/{} failed to get latest base token price, retrying in {} seconds...",
+                    attempts, max_retries, sleep_duration.as_secs()
+                );
+                sleep(sleep_duration).await;
+                retry_delay *= 2;
+                attempts += 1;
+            }
+        }
+        anyhow::bail!(
+            "Failed to get latest base token price after {} attempts",
+            max_retries
+        );
+    }
+
+    // Sleep for the remaining duration of the polling period
+    async fn sleep_until_next_fetch(&self, start_time: Instant) {
+        let elapsed_time = start_time.elapsed();
+        let sleep_duration = if elapsed_time >= self.config.price_polling_interval() {
+            Duration::from_secs(0)
+        } else {
+            self.config.price_polling_interval() - elapsed_time
+        };
+
+        tokio::time::sleep(sleep_duration).await;
+    }
+
+    /// If no base token L1 address is set then the base token is ETH.
+    fn is_base_token_eth(&self) -> bool {
+        self.base_token_l1_address.is_none()
+    }
+}
+
+#[async_trait]
+impl BaseTokenAdjuster for MainNodeBaseTokenAdjuster {
+    async fn convert(&self, params: FeeParams) -> anyhow::Result<FeeParams> {
+        let is_eth = self.is_base_token_eth();
+
+        if is_eth {
+            return Ok(params);
+        }
+
+        // Retries are necessary for the initial setup, where prices may not yet be persisted.
+        let latest_ratio = self.retry_get_latest_price().await?;
+
+        // TODO(PE-129): Implement latest ratio usability logic.
+
+        if let FeeParams::V2(params_v2) = params {
+            Ok(FeeParams::V2(convert_to_base_token(
+                params_v2,
+                latest_ratio,
+            )))
+        } else {
+            panic!("Custom base token is not supported for V1 fee model")
+        }
+    }
+}
+
+/// Converts the fee parameters to the base token using the latest ratio fetched from the DB.
+fn convert_to_base_token(params: FeeParamsV2, base_token_to_eth: BigDecimal) -> FeeParamsV2 {
+    let FeeParamsV2 {
+        config,
+        l1_gas_price,
+        l1_pubdata_price,
+    } = params;
+
+    let convert_price = |price_in_wei: u64| -> u64 {
+        let converted_price_bd = BigDecimal::from(price_in_wei) * base_token_to_eth.clone();
```

Comment:
the conversion rate should be `eth_to_base_token` ? 
according to doc: https://github.com/matter-labs/zksync-era/blob/627aab9703c47795247f8b6d21533520498ed025/docs/guides/advanced/fee_model.md#custom-base-token-configurations

Reply:
@XinyuCRO I'm not sure I'm following. The PR was in draft mode and this was changed. If this is still relevant please comment again.